src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,130 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		var releaseData = await UpdateChecker.GetLatestReleaseDataAsync();
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		var latestVersion = new Version( UpdateChecker.GetLatestReleaseVersion( releaseData ) );
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+		string archiveName = GetOSFileName();
+		string downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl
+			?? string.Empty;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string? downloadPath = Path.GetTempFileName();
+
+		string? currentProcessPath = Environment.ProcessPath;
+		string? currentProcessFileName = Path.GetFileNameWithoutExtension( currentProcessPath );
+		string backupPath = $"{currentProcessFileName}v{localVersion}.old.bak";
+
+		if( !string.IsNullOrEmpty( currentProcessPath ) ) {
+			File.Move( currentProcessPath, backupPath );
```

Comment:
Am getting an IO exception with permission denied message here on Mac. We should catch that and if so prompt the user to elevate permission

Also instead of creating the backup file in the existing directory we should maybe put it somewhere under .bmx. After the update happens and the user runs bmx again, it would try to clear out the backup file in /bin/ again which would then fail as it requires sudo to delete

Reply:
Just made some changes to prompt on IOexceptions! Should catch it but I don't have a mac to test